bluetooth: controller: Moved vendor specific HCI to hal

mtpr-ot opened the pull request:
Moved vendor specific part of hci.c to hal, thereby enabling out-of-tree
HCI implementations.

Signed-off-by: Morten Priess <mtpr@oticon.com>

carlescufi commented on subsys/bluetooth/controller/hci/nordic/hal/nrf5/hci.c:
```diff
@@ -0,0 +1,281 @@
+}
+#endif /* CONFIG_BT_HCI_VS_EXT */
+
+int vendor_cmd_handle(u16_t ocf, struct net_buf *cmd,
```
We should probably keep this generic demux function in the common code don't you think?

mtpr-ot replied:
@carlescufi: This is where the vendor adds own HCI command handlers, so I don't see how that would be possible?

mtpr-ot replied:
Perhaps we could have the original `vendor_cmd_handle` in common code with a different name. Then the vendor version of `vendor_cmd_handle` could call the common one, and ond only have own handers in the switch?

mtpr-ot replied:
Sorry, above comment about having a common `vendor_cmd_handle` would require moving all code back to common, and that really doesn't make sense to me. Please disregard that suggestion.

mtpr-ot committed:
bluetooth: controller: Moved vendor specific HCI to hal

Moved vendor specific part of hci.c to hal, thereby enabling out-of-tree
HCI implementations.

Signed-off-by: Morten Priess <mtpr@oticon.com>

joerchan reviewed (commented):
It seems like we are moving a lot of general code that would have to be duplicated.
Is it possible instead to define some hal functions for the vendors to implement.

Like:
hal_static_addr_get()
hal_ir_get()
hal_er_get()
hal_hw_version_get()
etc. ?
Otherwise it seems very strange to me to have HCI code implemented under HAL.

mtpr-ot commented:
> It seems like we are moving a lot of general code that would have to be duplicated.
> Is it possible instead to define some hal functions for the vendors to implement.
> 
> Like:
> hal_static_addr_get()
> hal_ir_get()
> hal_er_get()
> hal_hw_version_get()
> etc. ?
> Otherwise it seems very strange to me to have HCI code implemented under HAL.

@joerchan I understand your objections. The purpose of the change is twofold; 1) Allow vendor specific implementations of the functions you list, and 2) Enable creation of new vendor specific HCI commands, and have them handled in `vendor_cmd_handle`.

It would work for us to keep the exisiting implementation, just renaming `vendor_cmd_handle` to, for instance `vendor_cmd_handle_common`. Then vendor implements `vendor_cmd_handle`, calling `vendor_cmd_handle_common`. Then the vendor implementation is in control of which commands to map to other local functions, and support own HCI commands.

mtpr-ot commented:
Closing this PR. It has been replaced by #15661, which contains a much less intrusive change.
